src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -211,93 +211,82 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
 
-            if (!string.IsNullOrEmpty(tenant))
-            {
-                authority = _application.Authority.Replace(new Uri(_application.Authority).PathAndQuery, $"/{tenant}/");
-            }
-            else
-            {
-                authority = _application.Authority;
-            }
+            _application = await GetOrBuildConfidentialClientApplicationAsync().ConfigureAwait(false);
+            string authority = CreateAuthorityBasedOnTenantIfProvided(_application, tenant);
+            string? accessToken;
 
             try
             {
-                accessToken = await GetAccessTokenOnBehalfOfUserFromCacheAsync(
+                // Access token will return if call is from a web API
+                accessToken = await GetTokenForWebApiToCallDownstreamApiAsync(
                     _application,
-                    user,
-                    scopes,
                     authority,
-                    userFlow)
-                    .ConfigureAwait(false);
+                    scopes).ConfigureAwait(false);
+
+                if (!string.IsNullOrEmpty(accessToken))
+                {
+                    return accessToken;
+                }
+
+                // If access token is null, this is a web app
+                return await GetAccessTokenForWebAppWithAccountFromCacheAsync(
+                     _application,
+                     user,
+                     scopes,
+                     authority,
+                     userFlow)
+                     .ConfigureAwait(false);
             }
             catch (MsalUiRequiredException ex)
             {
-                // GetAccessTokenForUserAsync is an abstraction that can be called from a web app or a web API
+                // GetAccessTokenForUserAsync is an abstraction that can be called from a Web App or a Web API
                 _logger.LogInformation(ex.Message);
 
-                // to get a token for a Web API on behalf of the user, but not necessarily with the on behalf of OAuth2.0
-                // flow as this one only applies to Web APIs.
-                JwtSecurityToken? validatedToken = CurrentHttpContext?.GetTokenUsedToCallWebAPI();
+                // Case of the Web App: we let the MsalUiRequiredException be caught by the
+                // AuthorizeForScopesAttribute exception filter so that the user can consent, do 2FA, etc ...
+                throw new MicrosoftIdentityWebChallengeUserException(ex, scopes.ToArray());
+            }
+        }
+
+        private async Task<string?> GetTokenForWebApiToCallDownstreamApiAsync(
+            IConfidentialClientApplication application,
+            string authority,
+            IEnumerable<string> scopes)
+        {
+            try
+            {
+                // In web API, validatedToken will not be null
+                JwtSecurityToken? validatedToken = CurrentHttpContext.GetTokenUsedToCallWebAPI();
```

Comment:
```suggestion
                JwtSecurityToken? validatedToken = CurrentHttpContext?.GetTokenUsedToCallWebAPI();
```

Reply:
fixed. thanks

---
In reply to: [464810227](https://github.com/AzureAD/microsoft-identity-web/pull/383#discussion_r464810227) [](ancestors = 464810227)